### PR TITLE
 [sc-22645] Add Active Directory Adapter Catalog

### PR DESCRIPTION
**Pull request by Mohitp98**
#### This is an initial version of the adapter catalog: 

**Shortcut story**

[sc-22645](https://app.shortcut.com/sgnl/story/22645/sor-support-active-directory-as-a-system-of-record)

### CODE REVIEW

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:3]: Here is the b64 encoded SVG for icon

```
PHN2ZyB3aWR0aD0iNjQiIGhlaWdodD0iNjQiIHZpZXdCb3g9IjAgMCA2NCA2NCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHJlY3Qgd2lkdGg9IjY0IiBoZWlnaHQ9IjY0IiBmaWxsPSJ3aGl0ZSIvPgo8cGF0aCBkPSJNMjguOTAzMiAxMy4zNjg4TDEwLjIwMzIgMzQuNDI4QzguODI1NiAzNS45NzkyIDkuMTY5NiAzOC4zOTUyIDEwLjkyNCAzOS41MDE2TDMwLjEwMTYgNTEuNTkxMkMzMS4xOTY4IDUyLjI4MTYgMzIuNTkwNCA1Mi4yODE2IDMzLjY4NDggNTEuNTkxMkw1Mi44NjI0IDM5LjUwMTZDNTQuNjE2OCAzOC4zOTUyIDU0Ljk2IDM1Ljk3OTIgNTMuNTgzMiAzNC40MjhMMzQuODg0OCAxMy4zNjg4QzMzLjI5MzYgMTEuNTc2OCAzMC40OTQ0IDExLjU3NjggMjguOTAyNCAxMy4zNjg4SDI4LjkwMzJaIiBmaWxsPSJ1cmwoI3BhaW50MF9saW5lYXJfMTYyNV80MCkiLz4KPHBhdGggZD0iTTQyLjYxMzYgMzUuMDRMMjcuMzI4IDQ0Ljc2NzJDMjQuODQxNiA0Ni4zNDk2IDIxLjY2NDggNDYuMzUxMiAxOS4xNzYgNDQuNzcyOEwxMC44NCAzOS40ODMyQzkuMDg4MDEgMzguMzcxMiA4Ljc1MjAxIDM1Ljk1MjggMTAuMTM2IDM0LjQwNjRMMjguOTM4NCAxMy4yOTQ0QzMwLjQ2OTYgMTEuNTgyNCAzMy4zMTEyIDExLjU4ODggMzQuODM0NCAxMy4zMDcyTDQxLjI1MzYgMjAuNTZDNDEuMjUzNiAyMC41NiAzNi4xMzM2IDE3LjkyIDMxLjg5MzYgMjIuNzJMNDIuNjEzNiAzNS4wNFoiIGZpbGw9InVybCgjcGFpbnQxX2xpbmVhcl8xNjI1XzQwKSIvPgo8cGF0aCBkPSJNMzEuODkzNiAyMi43MkwyMS4xNzM2IDM1LjA0TDMxLjg5MzYgNDEuNzZMNDIuNjEzNiAzNS4wNEwzMS44OTM2IDIyLjcyWiIgZmlsbD0idXJsKCNwYWludDJfbGluZWFyXzE2MjVfNDApIi8+CjxtYXNrIGlkPSJtYXNrMF8xNjI1XzQwIiBzdHlsZT0ibWFzay10eXBlOmx1bWluYW5jZSIgbWFza1VuaXRzPSJ1c2VyU3BhY2VPblVzZSIgeD0iOSIgeT0iMTIiIHdpZHRoPSI0NiIgaGVpZ2h0PSI0MSI+CjxwYXRoIGQ9Ik0yOC45MDMyIDEzLjM2ODhMMTAuMjAzMiAzNC40MjhDOC44MjU2IDM1Ljk3OTIgOS4xNjk2IDM4LjM5NTIgMTAuOTI0IDM5LjUwMTZMMzAuMTAxNiA1MS41OTEyQzMxLjE5NjggNTIuMjgxNiAzMi41OTA0IDUyLjI4MTYgMzMuNjg0OCA1MS41OTEyTDUyLjg2MjQgMzkuNTAxNkM1NC42MTY4IDM4LjM5NTIgNTQuOTYgMzUuOTc5MiA1My41ODMyIDM0LjQyOEwzNC44ODQ4IDEzLjM2ODhDMzMuMjkzNiAxMS41NzY4IDMwLjQ5NDQgMTEuNTc2OCAyOC45MDI0IDEzLjM2ODhIMjguOTAzMloiIGZpbGw9IndoaXRlIi8+CjwvbWFzaz4KPGcgbWFzaz0idXJsKCNtYXNrMF8xNjI1XzQwKSI+CjxwYXRoIHN0eWxlPSJtaXgtYmxlbmQtbW9kZTptdWx0aXBseSIgZD0iTTMxLjg5MzYgNTIuNzJWOS45Mkw1Ni4yMTM2IDM3LjQ0TDMxLjg5MzYgNTIuNzJaIiBmaWxsPSJ1cmwoI3BhaW50M19saW5lYXJfMTYyNV80MCkiLz4KPC9nPgo8ZGVmcz4KPGxpbmVhckdyYWRpZW50IGlkPSJwYWludDBfbGluZWFyXzE2MjVfNDAiIHgxPSI5LjM1NTIiIHkxPSIzMi4wNjY0IiB4Mj0iNTQuNDMyIiB5Mj0iMzIuMDY2NCIgZ3JhZGllbnRVbml0cz0idXNlclNwYWNlT25Vc2UiPgo8c3RvcCBzdG9wLWNvbG9yPSIjMjI1MDg3Ii8+CjxzdG9wIG9mZnNldD0iMSIgc3RvcC1jb2xvcj0iIzAzNEVBRiIvPgo8L2xpbmVhckdyYWRpZW50Pgo8bGluZWFyR3JhZGllbnQgaWQ9InBhaW50MV9saW5lYXJfMTYyNV80MCIgeDE9IjkuMjgwMDEiIHkxPSIyOC45ODQ4IiB4Mj0iNDIuNjEzNiIgeTI9IjI4Ljk4NDgiIGdyYWRpZW50VW5pdHM9InVzZXJTcGFjZU9uVXNlIj4KPHN0b3Agc3RvcC1jb2xvcj0iIzY5REVGRiIvPgo8c3RvcCBvZmZzZXQ9IjEiIHN0b3AtY29sb3I9IiMxREE2RTgiLz4KPC9saW5lYXJHcmFkaWVudD4KPGxpbmVhckdyYWRpZW50IGlkPSJwYWludDJfbGluZWFyXzE2MjVfNDAiIHgxPSIzMS44OTM2IiB5MT0iMjIuNzIiIHgyPSIzMS44OTM2IiB5Mj0iNDEuNzYiIGdyYWRpZW50VW5pdHM9InVzZXJTcGFjZU9uVXNlIj4KPHN0b3Agc3RvcC1jb2xvcj0iIzhDRUFGRiIvPgo8c3RvcCBvZmZzZXQ9IjEiIHN0b3AtY29sb3I9IiM1RERGRkEiLz4KPC9saW5lYXJHcmFkaWVudD4KPGxpbmVhckdyYWRpZW50IGlkPSJwYWludDNfbGluZWFyXzE2MjVfNDAiIHgxPSIyNS43ODg4IiB5MT0iMTMuNDQ0OCIgeDI9IjQ0LjMyMjQiIHkyPSI0NS41NDQ4IiBncmFkaWVudFVuaXRzPSJ1c2VyU3BhY2VPblVzZSI+CjxzdG9wIHN0b3Atb3BhY2l0eT0iMC4xIi8+CjxzdG9wIG9mZnNldD0iMSIgc3RvcC1vcGFjaXR5PSIwLjIiLz4KPC9saW5lYXJHcmFkaWVudD4KPC9kZWZzPgo8L3N2Zz4=
```

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:21]:
```suggestion
adapterConfig: "ewogICAgImJhc2VETiI6ICJ7e0lucHV0IFJlcXVpcmVkOiBCYXNlRE59fSIsCiAgICAiY2VydGlmaWNhdGVDaGFpbiI6ICJ7e0lucHV0IFJlcXVpcmVkOiBJZiBMREFQUyBpcyBlbmFibGVkfX0iLAogICAgImZpbHRlcnMiOiB7CiAgICAgICAgIlVzZXIiOiAiKCYob2JqZWN0Q2xhc3M9dXNlcikpIiwKICAgICAgICAiR3JvdXAiOiAiKCYob2JqZWN0Q2xhc3M9Z3JvdXApKSIKICAgIH0KfQ=="
```

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:32]:
```suggestion
    pageSize: 100
```

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:37]:
```suggestion
        description: The Distinguished Name (DN) of the user
```

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:42]:
```suggestion
      - name: objectClass
        externalId: objectClass
```

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:52]: Maybe index this

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:55]: Is this an integer?

Also, please paste the link to AD schema docs on the top of this file for easy reference.

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:126]:
```suggestion
    pageSize: 100
```

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:131]:
```suggestion
        description: The Distinguished Name (DN) of the group
```

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:123]: Please follow the format of "An entity representing a XYZ in an Active Directory" for all entities

Please refer to latest catalogs if needed.

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:181]:
```suggestion
    pageSize: 100
```

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:202]: nit: You can use `attributeAlias` to have more meaningful names if attributeExternalId is unclear

**Review by r-rakshith — CHANGES_REQUESTED** [review]
Looks good overall

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [reply activedirectory.sgnl.yaml:21]: The default config must be production ready. Add `{{Input Required}}` placeholders wherever you expect user configurable input.

**Comment by r-rakshith** [comment]
Tagging @marc-sgnl to identify a minimal set of attributes for this SoR. Not a blocker for this PR as this can be done in a follow up PRs as well.

File: activedirectory.sgnl.yaml
Comment (Mohitp98) [reply activedirectory.sgnl.yaml:52]: `dn` is correct indexed attribute. A distinguished name (usually just shortened to “DN”) uniquely identifies an entry and describes its position in the DIT. We are using `dn` as `uniqueIDAttribute` for Entity in the adapter.

Ref: https://ldap.com/ldap-dns-and-rdns/

**Comment by marc-sgnl** [comment]
> Tagging @marc-sgnl to identify a minimal set of attributes for this SoR. Not a blocker for this PR as this can be done in a follow up PRs as well.

Ack np - I'll wait until we merge this and then create a PR

File: activedirectory.sgnl.yaml
Comment (Mohitp98) [reply activedirectory.sgnl.yaml:55]: Yes, Please check this: https://learn.microsoft.com/en-us/windows/win32/adschema/a-instancetype


Added reference to AD Schema

**Comment by Mohitp98** [comment]
I feel we should create a catalog per distribution, It'll be easier to maintain the catalog going ahead.

WDYT @r-rakshith?

**Review by marc-sgnl — COMMENTED** [review]
Couple of comments

File: activedirectory.sgnl.yaml
Comment (Mohitp98) [reply activedirectory.sgnl.yaml:37]: In this case, we have to update `uniqueIDAttribute` in the adapter implementation from `dn` to `objectGUID`. This changes can be added as part of on-going phase 2 changes of adapter.

CC: @r-rakshith

File: activedirectory.sgnl.yaml
Comment (Mohitp98) [reply activedirectory.sgnl.yaml:132]: @marc-sgnl Follow-up question:

###  Question: What should be the format of `objectGUID` and `objectSid` ?

Currently, `ldapsearch` returns the base64 encoded value of these IDs wherein
- `objectGUID` is a `UUID`(*16-byte structure*)
- `objectSid` is a `String(Sid)` (*An octet string that contains a security identifier*)

**check ref for more details**

```
# Result from ldapSearch

# marc, Users, corp.sgnl.ai
dn: CN=marc,CN=Users,DC=corp,DC=sgnl,DC=ai
objectClass: top
objectClass: person
objectClass: organizationalPerson
objectClass: user
cn: marc
description: Built-in account for administering the computer/domain
distinguishedName: CN=marc,CN=Users,DC=corp,DC=sgnl,DC=ai
instanceType: 4
whenCreated: 20231013202314.0Z
whenChanged: 20240428061131.0Z
uSNCreated: 8196
memberOf: CN=ADSyncAdmins,CN=Users,DC=corp,DC=sgnl,DC=ai
memberOf: CN=Group Policy Creator Owners,CN=Users,DC=corp,DC=sgnl,DC=ai
memberOf: CN=Domain Admins,CN=Users,DC=corp,DC=sgnl,DC=ai
memberOf: CN=Enterprise Admins,CN=Users,DC=corp,DC=sgnl,DC=ai
memberOf: CN=Schema Admins,CN=Users,DC=corp,DC=sgnl,DC=ai
memberOf: CN=Administrators,CN=Builtin,DC=corp,DC=sgnl,DC=ai
name: marc
objectGUID:: /EV8bsrijE+nd5JP+9Zl9w==
objectSid:: AQUAAAAAAAUVAAAAxpq11ZhA0RW0bN0o9AEAAA==
objectCategory: CN=Person,CN=Schema,CN=Configuration,DC=corp,DC=sgnl,DC=ai
```

### Actual values:
```
objectGUID: fc457c6e-cae2-8c4f-a777-924ffbd665f7
objectSid: S-1-5-21-3585448646-366035096-685599924-500

```


Ref: 
- [objectGUID attribute](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-ada3/937eb5c6-f6b3-4652-a276-5d6bb8979658)
- [Object-Sid attribute](https://learn.microsoft.com/en-us/windows/win32/adschema/a-objectsid)

Comment (marc-sgnl) [reply activedirectory.sgnl.yaml:132]: Thanks @Mohitp98  -- We'd expect their Actual Values, this is what AD Admins in particular would be building on top of. The Base64 encoding is a nuance of the ldapsearch tool.

Comment (r-rakshith) [reply activedirectory.sgnl.yaml:132]: @Mohitp98, can you please confirm if the adapter response from the go library we use contains the actual values and not encoded values?

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [reply activedirectory.sgnl.yaml:37]: With the generic adapter, one should be able to make these changes via the catalog with no changes to the adapter's codebase.

Comment (marc-sgnl) [reply activedirectory.sgnl.yaml:37]: > With the generic adapter, one should be able to make these changes via the catalog with no changes to the adapter's codebase.

Agree - but our customers shouldn't use this as the uniqueID given it's not durable, so the sooner we can get this fixed in our default config the better.

Comment (r-rakshith) [reply activedirectory.sgnl.yaml:37]: +1

File: activedirectory.sgnl.yaml
Comment (Mohitp98) [reply activedirectory.sgnl.yaml:132]: Using `ldap-go`, [EntryAttributes](https://pkg.go.dev/github.com/go-ldap/ldap/v3#EntryAttribute) will return value as a [[]string](https://pkg.go.dev/builtin#string) or [[][]byte](https://pkg.go.dev/builtin#byte).

In case of **objectGUID** & **objectSid**, using `[]string` as it is won't be helpful as it is a simple string conversion of actual []byte data. This **does not represent** the actual formatted string.

Example: 

```
objectGUID:"\xfcE|n\xca\xe2\x8cO\xa7w\x92O\xfb\xd6e\xf7"
ojbectSid: "\x01\x05\x00\x00\x00\x00\x00\x05\x15\x00\x00\x00ƚ\xb5\u0558@\xd1\x15\xb4l\xdd(\xf4\x01\x00\x00"
```

To retrieve the actual values of these attributes, we have to handle this using appropriate method:
- ObjectGUID: we can use [ParseBytes](https://pkg.go.dev/github.com/google/uuid#ParseBytes) from github.com/google/uuid
- ObjectSid: We can use [Decode](https://pkg.go.dev/github.com/bwmarrin/go-objectsid@v0.0.0-20191126144531-5fee401a2f37#Decode) from github.com/bwmarrin/go-objectsid

File: activedirectory.sgnl.yaml
Comment (Mohitp98) [reply activedirectory.sgnl.yaml:37]: JFI, As we are using OpenLDAP version for testing the adapter, we are not able to see `objectGUID` or `objectSid` for any of the entities.

Sample output from ldapSearch command:
```
❯ ldapsearch -x -b "dc=example,dc=org" -H ldap://localhost:32828 -D "cn=admin,dc=example,dc=org" -w "admin" "(&(objectClass=inetOrgPerson)(objectClass=person))"

# alice, People, example.org
dn: cn=alice,ou=People,dc=example,dc=org
objectClass: person
objectClass: inetOrgPerson
objectClass: organizationalPerson
objectClass: top
cn: alice
userPassword:: cGFzcw==
givenName: Alice
sn: lorem
mail: alice@oncase.com.br
uid: 1004

# Barbara Jensen, People, example.org
dn: cn=Barbara Jensen,ou=People,dc=example,dc=org
objectClass: person
objectClass: inetOrgPerson
objectClass: organizationalPerson
objectClass: top
cn: Barbara Jensen
cn: Babs Jensen
displayName: Babs Jensen
sn: Jensen
givenName: Barbara
initials: BJJ
title: manager, product development
uid: 1007
mail: bjensen@siroe.com
telephoneNumber: +1 408 555 1862
facsimileTelephoneNumber: +1 408 555 1992
mobile: +1 408 555 1941
mobile: +1 408 555 2014
roomNumber: 0209
carLicense: 6ABC246
o: Siroe
ou: Product Development
departmentNumber: 2604
employeeNumber: 42
employeeType: full time
preferredLanguage: fr, en-gb;q=0.8, en;q=0.7
labeledURI: http://www.siroe.com/users/bjensen My Home Page
```

Thus changing `uniqueIDAttribute` will not work for `OpenLDAP` distribution.

Comment (Mohitp98) [reply activedirectory.sgnl.yaml:37]: For future reference,

As `uniqueIdAttribute` can vary with server implementation.

We decided to set `uniqueIdAttribute` for a request from [AttributeConfig.UniqueId](https://github.com/SGNL-ai/adapter-framework/blob/main/adapter.go#L120).

Closing this conversation now.

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:34]: As mentioned in [previous comments](https://github.com/SGNL-ai/catalog-sgnl-templates/pull/56#discussion_r1588406111), please add correct placeholders for `baseDN`, `certificateChain` etc.

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:349]: As discussed previously, `objectGUID` would be misleading here as this attribute is a composite-id generated from two other attributes. Please take care while naming attributes to prevent any confusion while writing policies.
```suggestion
      - name: id
        externalId: id
        type: String
        indexed: true
        uniqueId: true
```

**Review by r-rakshith — CHANGES_REQUESTED** [review]

File: activedirectory.sgnl.yaml
Comment (Mohitp98) [reply activedirectory.sgnl.yaml:34]:
```
{
    "baseDN": "{{Input Required: BaseDN}}",
    "entityConfig": {
        "User": {
            "selector": "(&(objectCategory=user)(objectClass=user)(distinguishedName=*))"
        },
        "Computer": {
            "selector": "(&(objectCategory=computer)(name=*))"
        },
        "Group": {
            "selector": "(&(objectCategory=group)(objectClass=group)(distinguishedName=*))"
        },
        "GroupMember": {
            "memberOf": "Group",
            "selector": "(&(memberOf={{CollectionId}})(objectCategory=user)(objectClass=user))",
            "memberUniqueIdAttribute": "{{Input Required: If memberOf is present}}",
            "memberOfUniqueIdAttribute": "{{Input Required: If memberOf is present}}"
        }
    },
    "certificateChain": "{{Input Required: If LDAPS is enabled: base64 encoded certificateChain}}"
}
```

Updated default config.

Comment (marc-sgnl) [reply activedirectory.sgnl.yaml:34]: Can we replace `selector` here with `query` -- we should seem semantically familiar to folks who are administrators of AD and LDAP systems, and in these fields you enter an LDAP Query

Comment (marc-sgnl) [reply activedirectory.sgnl.yaml:34]: Tangentially, @r-rakshith @Mohitp98  I worry we've preemptively generalized here. For Active Directory, we know what `memberUniqueIdAttribute` and `memberOfUniqueIdAttribute` are going to be. At a minimum, we should pre-populate these for an SoR Admin with the appropriate values for AD, but we should chat about further iterations where we remove this setting completely and instead replace it with some higher level setting that defines the type of system we're connecting to, e.g. `"Type": "ActiveDirectory"`. Things like AD are going to have a range of specificities that I don't think we should force an administrator to configure and manage.

Comment (r-rakshith) [reply activedirectory.sgnl.yaml:34]: That sounds good to me. I was thinking of the same given the oddities in Active Directory. We might end up having multiple implementations of the adapter although they're technically LDAP. We should chat more about this!


In the meantime, 
This is what the current configuration example looks like. The initial idea was to use a combination of "selector" to select an entity and "filter" to get only a subset of those.
I have no objections reducing all of that to "query" if that is more suitable for administrators. Let me know what you think/
```json
{
    "baseDN": "dc=org,dc=example,dc=io",
    "certificateChain": "....",
    "entityConfig": {
        "User": {
            "selector": "(&(objectCategory=user)(objectClass=user)(distinguishedName=*))",
            "filter": "(&(logonCount=0))"
        },
        "Group": {
            "selector": "(&(objectCategory=group)(objectClass=group)(distinguishedName=*))",
            "filter": "(&(instanceType=4)(sAMAccountName=Blocked Sign-In))"
        },
        "GroupMember": {
            "memberOf": "Group",
            "selector": "(&(memberOf={{CollectionId}}))",
			"filter": "(&(logonCount=0))",
			"memberUniqueIdAttribute": "memberDistinguishedName",
			"memberOfUniqueIdAttribute": "groupDistinguishedName"
        }
    }
}
```

Comment (Mohitp98) [reply activedirectory.sgnl.yaml]: @marc-sgnl , @r-rakshith  Here is the updated config:

Reduced `filter` & `selector` attributes to a single `query` attribute.

```json
{
    "baseDN": "{{Input Required: BaseDN}}",
    "certificateChain": "{{Input Required: If LDAPS is enabled: base64 encoded certificateChain}}",
    "entityConfig": {
        "User": {
            "query": "(&(objectCategory=user)(objectClass=user)(distinguishedName=*)(logonCount=0))"
        },
        "Computer": {
            "query": "(&(objectCategory=computer)(name=*))"
        },
        "Group": {
            "query": "(&(objectCategory=group)(objectClass=group)(distinguishedName=*)(instanceType=4))"
        },
        "GroupMember": {
            "memberOf": "Group",
            "query": "(&(memberOf={{CollectionId}})(objectCategory=user)(objectClass=user))",
            "memberUniqueIdAttribute": "{{Input Required: If memberOf is present}}",
            "memberOfUniqueIdAttribute": "{{Input Required: If memberOf is present}}"
        }
    }
}
```

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:35]: Let's populate default and valid values for 
- `memberUniqueIdAttribute` and 
- `memberOfUniqueIdAttribute`

The default adapter config in the template should be complete as much as possible to minimize user effort

Comment (r-rakshith) [reply activedirectory.sgnl.yaml]: On decoding the default adapterConfig, I see GroupMember configured as
```json
"GroupMember": {
            "memberOf": "Group",
            "query": "(&(memberOf={{CollectionId}})(objectCategory=user)(objectClass=user))",
            "memberUniqueIdAttribute": "memberObjectGUID",
            "memberOfUniqueIdAttribute": "groupObjectGUID"
        }
```

Neither `memberObjectGUID` nor `groupObjectGUID` is present as attributes on  user object fetched via `objectClass=user` and `objectClass=group`. I see `objectGUID` as an attribute on both these entities. What am I missing here?

File: activedirectory.sgnl.yaml
Comment (Mohitp98) [reply activedirectory.sgnl.yaml:349]: To avoid hard-coded attribute, this will be targeted in improvements for adapter, as generic attributes are not currently supported

File: activedirectory.sgnl.yaml
Comment (Mohitp98) [reply activedirectory.sgnl.yaml]: This will be targeted in improvements for adapter, as generic attributes are not currently supported. We'll need the capability in adapter to parse attributes using json path format similar to what we are currently using in other adapters.

File: activedirectory.sgnl.yaml
Comment (r-rakshith) [inline activedirectory.sgnl.yaml:56]: This is fine for an `activedirectory` template. 

Please create an `ldap` template that contains detailed documentation on setting up `entityConfig` to ingest any kind of built-in/custom entity of various types (i.e. a standalone entity vs a member entity). Not a blocker for this PR.

**Review by r-rakshith — APPROVED** [review]

File: activedirectory.sgnl.yaml
Comment (Mohitp98) [reply activedirectory.sgnl.yaml:56]: Added a task for `ldap` template: [#25545](https://app.shortcut.com/sgnl/story/25545/sor-improvements-for-active-directory-adapter)

**Pull request merged**
